meeteval/io/seglst.py

```diff
@@ -0,0 +1,388 @@
+        """
+        A segment.
+
+        Note:
+            We do not define an enum with all these keys for speed reasons
+        """
+        session_id: str
+        start_time: float
+        end_time: float
+        words: str
+        speaker: str
+        segment_index: int
+
+        # Unused but by MeetEval but present in some file formats. They are defined here for compatibility and
+        # conversion in both directions
+        channel: int
+        confidence: float
+except ImportError:
+    # Python <3.8 doesn't have TypedDict
+    SegLstSegment = typing.Dict[str, typing.Any]
+
+
+@dataclasses.dataclass(frozen=True)
+class SegLST(BaseABC):
+    """
+    A collection of segments in SegLST format. This the input type to most
+    functions in MeetEval that process transcript segments.
+    """
+    segments: 'List[SegLstSegment]'
+
+    # Caches
+    _unique = None
+
+    @cached_property
+    def keys(self):
+        """
+        Keys that are common among all segments
+        """
+        if not self.segments:
+            return set()
+        return set.intersection(*[set(s.keys()) for s in self])
+
+    def unique(self, key):
+        """
+        Returns the unique values for `key` among all segments.
+        """
+        return set([s[key] for s in self.segments])
+
+    def __iter__(self):
+        return iter(self.segments)
+
+    def __getitem__(self, item):
+        return self.segments[item]
+
+    def __len__(self):
+        return len(self.segments)
+
+    def __add__(self, other):
+        if isinstance(other, SegLST):
+            return SegLST(self.segments + other.segments)
+        return NotImplemented
+
+    def groupby(self, key):
+        """
+        >>> t = asseglst(['a b c', 'd e f', 'g h i'])
+        >>> t.segments
+        [{'words': 'a b c', 'segment_index': 0, 'speaker': 0}, {'words': 'd e f', 'segment_index': 0, 'speaker': 1}, {'words': 'g h i', 'segment_index': 0, 'speaker': 2}]
+
+        >>> from pprint import pprint
+        >>> pprint(t.groupby('speaker')) # doctest: +ELLIPSIS
+        {0: SegLST(segments=[{'words': 'a b c', 'segment_index': 0, 'speaker': 0}]),
+         1: SegLST(segments=[{'words': 'd e f', 'segment_index': 0, 'speaker': 1}]),
+         2: SegLST(segments=[{'words': 'g h i', 'segment_index': 0, 'speaker': 2}])}
+        """
+        return {k: SegLST(g) for k, g in groupby(self.segments, key=key).items()}
+
+    def sorted(self, key):
+        """
+        Returns a copy of this object with the segments sorted by `key`.
+        """
+        return SegLST(sorted(self.segments, key=_get_key(key)))
+
+    def map(self, fn):
+        """
+        Applies `fn` to all segments and returns a new `SegLST` object with the results.
+        """
+        return SegLST([fn(s) for s in self.segments])
+
+    def flatmap(self, fn):
+        """
+        Applies `fn` to all segments, flattens the output and returns a new `SegLST` object with the results.
```

Comment:
I thought it was a well-known and commonly used operation. I'll update the docstring to be clearer about what the function is doing and give a simple example for splitting utterances into words:

```python
s = SegLST([{'words': 'a b'}, {'words': 'c d'}])
s.flatmap(lambda x: [{'words': w} for w in x['words'].split()])
SegLST(segments=[{'words': 'a'}, {'words': 'b'}, {'words': 'c'}, {'words': 'd'}])
```

Reply:
It is rarely used in python. I usually see list comprehensions, where a flatmap could be used.
The code is longer, but easier to modify (i.e. not restricted to the flatmap operation).
```python
SegLST([
    entry
    for x in s
    for entry in [{'words': w} for w in x['words'].split()]
])
```

Reply:
How about this?

```python
def apply(self, fn):
    return SegLST(fn(self.segments))
```

Would result in 

```python
s.apply(lambda s: [
    entry
    for x in s
    for entry in [{'words': w} for w in x['words'].split()]
])
```

Now that I look at it, it doesn't improve over the list-comprehension variant. (Python syntax for this kind of thing sucks)

Reply:
Yes, working with wrappers around list, it is annoying and error-prone to work with list comprehensions.

I like the `flatmap` code more. It solves this particular problem better and the name is well-defined by java.

Since I am more familiar with data loading pipelines, the description as map followed by unbatch is more familiar to me.

The `apply` code gets shorter, when you don't use an `apply` function:
```python
s.__class__([
    entry
    for x in s
    for entry in [{'words': w} for w in x['words'].split()]
])
```

Reply:
Updated the type annotations and docstring to make clearer, what `flatmap` does